pkg/smart/parse.go

```diff
@@ -0,0 +1,329 @@
+	InSmartctlDatabase bool    `json:"in_smartctl_database"`
+
+	SmartStatus *struct {
+		Passed bool `json:"passed"`
+	} `json:"smart_status"`
+
+	Temperature *struct {
+		Current float32 `json:"current"`
+	} `json:"temperature"`
+
+	PowerCycleCount *int `json:"power_cycle_count"`
+	PowerOnTime     *struct {
+		Hours int `json:"hours"`
+	} `json:"power_on_time"`
+
+	RotationRate      *string `json:"rotation_rate,omitempty"`
+	LogicalBlockSize  *int    `json:"logical_block_size,omitempty"`
+	PhysicalBlockSize *int    `json:"physical_block_size,omitempty"`
+
+	UserCapacity *struct {
+		Blocks int   `json:"blocks"`
+		Bytes  int64 `json:"bytes"`
+	} `json:"user_capacity"`
+
+	ATAVersion *struct {
+		String     string `json:"string"`
+		MajorValue int    `json:"major_value"`
+		MinorValue int    `json:"minor_value"`
+	} `json:"ata_version"`
+
+	SATAVersion *struct {
+		String string `json:"string"`
+		Value  int    `json:"value"`
+	} `json:"sata_version"`
+
+	InterfaceSpeed *struct {
+		Max struct {
+			SataValue      int    `json:"sata_value"`
+			String         string `json:"string"`
+			UnitsPerSecond int64  `json:"units_per_second"`
+			BitsPerUnit    int64  `json:"bits_per_unit"`
+		} `json:"max"`
+	} `json:"interface_speed"`
+
+	NVMESmartHealthInformationLog *nvmeSmartHealthInformationLog `json:"nvme_smart_health_information_log"`
+
+	ATASmartData *struct {
+		SelfTest struct {
+			String string `json:"string"`
+			Passed bool   `json:"passed"`
+		} `json:"self_test"`
+		Capabilities struct {
+			ExecOfflineImmediateSupport bool `json:"exec_offline_immediate_supported"`
+			OfflineIsAbortedUponNewCmd  bool `json:"offline_is_aborted_upon_new_cmd"`
+			OfflineSurfaceScanSupported bool `json:"offline_surface_scan_supported"`
+			SelfTestsSupported          bool `json:"self_tests_supported"`
+			ConveyanceSelfTestSupported bool `json:"conveyance_self_test_supported"`
+			SelectiveSelfTestSupported  bool `json:"selective_self_test_supported"`
+			AttributeAutoSaveEnabled    bool `json:"attribute_autosave_enabled"`
+			ErrorLoggingSupported       bool `json:"error_logging_supported"`
+			GpLoggingSupported          bool `json:"gp_logging_supported"`
+		}
+	} `json:"ata_smart_data"`
+
+	ATASmartAttributes *ataSmartAttributes `json:"ata_smart_attributes"`
+
+	ATASmartSelectiveSelfTestLog *struct {
+		Revision int `json:"revision"`
+
+		Flags struct {
+			ReminderScanEnabled bool `json:"reminder_scan_enabled"`
+		} `json:"flags"`
+		PowerUpScanResumeMinutes int `json:"power_up_scan_resume_minutes"`
+	} `json:"ata_smart_selective_self_test_log"`
+}
+
+var smartctlVersionRegexp *regexp.Regexp
+
+func init() {
+	smartctlVersionRegexp = regexp.MustCompile(`^smartctl\s(\d.\d)\s(\w|\W)+$`)
+}
+
+func DetectTools() error {
+	buildStr, err := checkTools()
+	if err != nil {
+		return err
+	}
+
+	return smartctlIsSupportedVersion(buildStr)
+}
+
+// Parse detect hardware disks and parse their S.M.A.R.T
+func Parse() (interface{}, []error) {
```

Comment:
Returning a strict type instead of interface{} (which is actually map[string]) can make the code more readable and consistent with other packages: almost all of them use struct definitions with `json` marshaling settings for returning result.

Reply:
Please look to answer above related to `AddInnerWithPrefix`

Reply:
This concern was raised already in the past, that we should do a proper re-factoring to be able to use `MeasurementMap` instead `interface{}`.
Guess for now we can't do much :/

Reply:
https://github.com/cloudradar-monitoring/cagent/pull/136

Reply:
done